iceoryx_hoofs/test/moduletests/error_reporting/test_error_reporting_api.cpp

```diff
@@ -0,0 +1,266 @@
+// Copyright (c) 2023 by Apex.AI Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+#include "test.hpp"
+#include <gtest/gtest.h>
+
+// some dummy modules under test
+#include "module_a/error_reporting.hpp"
+#include "module_b/error_reporting.hpp"
+
+// simplifies checking for errors during test execution
+#include "iceoryx_hoofs/testing/error_reporting/test_support.hpp"
+
+#include <iostream>
+
+namespace
+{
+
+// NOLINTBEGIN(cppcoreguidelines-avoid-do-while) bad rule, disable globally
+using namespace ::testing;
+using namespace iox::err;
+using namespace iox::cxx;
+using namespace iox::testing;
+
+using MyErrorA = module_a::errors::Error;
+using MyCodeA = module_a::errors::Code;
+
+using MyErrorB = module_b::errors::Error;
+using MyCodeB = module_b::errors::Code;
+
+class ErrorReportingApi_test : public Test
+{
+  public:
+    void SetUp() override
+    {
+        /// @todo iox-#1032 this should be done for all tests (using Gtest hooks) in integration
+        iox::testing::ErrorHandler::instance().reset();
+    }
+
+    void TearDown() override
+    {
+    }
+};
+
+TEST_F(ErrorReportingApi_test, panicWithoutMessage)
+{
+    ::testing::Test::RecordProperty("TEST_ID", "a55f00f1-c89d-4d4d-90ea-6ca510ad3942");
+    auto f = []() { IOX_PANIC(""); };
+
+    runInTestThread(f);
+
+    EXPECT_PANIC();
+}
+
+TEST_F(ErrorReportingApi_test, panicWithMessage)
+{
+    ::testing::Test::RecordProperty("TEST_ID", "cfbaf43b-de11-4858-ab86-ae3ae3fac2fe");
+    auto f = []() { IOX_PANIC("message"); };
+
+    runInTestThread(f);
+
+    EXPECT_PANIC();
+}
+
+TEST_F(ErrorReportingApi_test, reportNonFatal)
```

Comment:
Added for now. After some discussion with @elBoberido I came to the conclusion that we do not want to allow this to work (due to `IOX_REPORT_FATAL`) and subsequent return statements that must not exist in the fatal case (but have to in non fatal cases). 

I am of the opinion that if it has to be fatal `IOX_REPORT_FATAL` should be used and no subsequent return statement is needed. If only `IOX_REPORT` is used, a subsequent return statement is required regardless of whether the error is fatal or not. 

However, this can break code if an error kind changes from fatal to non-fatal if this is not enforced (which cannot be  done).

I prefer that it would work and hence added the test but I will think about it. I think it is not immediately harmful either way.